hal/src/electron/modem/dns_client.cpp

```diff
@@ -147,28 +153,58 @@ class UdpSocket {
             return MDM_SOCKET_ERROR;
         }
         sock_ = sock;
+        oldModemStateChangeCount_ = newModemStateChangeCount_;
         return 0;
     }
 
     void close() {
-        if (sock_ >= 0) {
+        if (_isSocketValid() && sock_ >= 0) {
             electronMDM.socketFree(sock_);
-            sock_ = -1;
+            sock_ = MDM_SOCKET_ERROR;
         }
     }
 
     int sendTo(MDM_IP addr, int port, const char* data, size_t size) {
+        if (!_isSocketValid()) {
+            // do not try to free socket since it is invalidated
+            return MDM_SOCKET_ERROR;
+        }
         return electronMDM.socketSendTo(sock_, addr, port, data, size);
     }
 
     int recvFrom(MDM_IP* addr, int* port, char* data, size_t size) {
+        if (!_isSocketValid()) {
+            // do not try to free socket since it is invalidated
+            return MDM_SOCKET_ERROR;
+        }
+        // See if any data is waiting for us
+        int result = electronMDM.socketReadable(sock_);
+        if (result <= 0) {
+            return 0; // no data
+        }
+
         return electronMDM.socketRecvFrom(sock_, addr, port, data, size);
     }
 
 private:
+    bool _isSocketValid() {
```

Comment:
yeah I struggled with that, because then if we use your name suggestion it doesn't make as much sense in other parts of the code:
```c++
    void close() {
        if (_isSocketValid() && sock_ >= 0) {
            electronMDM.socketFree(sock_);
            sock_ = MDM_SOCKET_ERROR;
        }
    }
```

I'll think about how to resolve this..

Reply:
I'm going to use your name, and invert the `bool` return values.  I think that makes it much more tidy, thanks!

Reply:
I typically make such methods return void, as the validity of an object in question usually can be checked by other means as well:
```
void close() {
    invalidateSocketIfNeeded();
    if (sock_ >= 0) {
        electronMDM.socketFree(sock_);
        sock_ = MDM_SOCKET_ERROR;
    }
}
```
Anyway, I'm just nitpicking here :)

Reply:
Yeah I was trying not to duplicate `sock_ = MDM_SOCKET_ERROR;` and tests for `sock_ >= 0` in multiple places.